flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java

```diff
@@ -633,17 +635,80 @@ static Resource getUnitResource(YarnConfiguration yarnConfig) {
+            throws IOException {
+        final List<String> providedDirs =
+                ConfigUtils.decodeListFromConfig(
+                        configuration, YarnConfigOptions.PROVIDED_LIB_DIRS, String::toString);
+        return getQualifiedRemoteSharedPaths(
+                providedDirs, YarnConfigOptions.PROVIDED_LIB_DIRS.key(), yarnConfiguration);
+    }
+
+    public static boolean isUsrLibIncludedInPath(final FileSystem fileSystem, final Path path)
+            throws IOException {
+        final FileStatus fileStatus = fileSystem.getFileStatus(path);
+        // Use the Path obj from fileStatus to get rid of trailing slash
+        return fileStatus.isDirectory()
+                && ConfigConstants.DEFAULT_FLINK_USR_LIB_DIR.equals(fileStatus.getPath().getName());
+    }
+
+    public static Path getQualifiedRemoteProvidedUsrLib(
+            org.apache.flink.configuration.Configuration configuration,
+            YarnConfiguration yarnConfiguration)
+            throws IOException, IllegalArgumentException {
+        String usrlib = configuration.getString(YarnConfigOptions.PROVIDED_USRLIB_DIR);
+        if (usrlib == null) {
+            return null;
+        }
+        final List<String> providedDirs = Collections.singletonList(usrlib);
+        final List<Path> paths =
+                getQualifiedRemoteSharedPaths(
+                        providedDirs,
+                        YarnConfigOptions.PROVIDED_USRLIB_DIR.key(),
+                        yarnConfiguration);
+        if (paths == null || paths.size() == 0) {
+            return null;
+        }
+        final Path providedUsrLibPath = paths.get(0);
+        checkArgument(
+                isUsrLibIncludedInPath(
+                                providedUsrLibPath.getFileSystem(yarnConfiguration),
+                                providedUsrLibPath)
+                        && isRemotePath(providedUsrLibPath.toString()),
+                "The %s must point to a remote dir named with %s.",
+                YarnConfigOptions.PROVIDED_USRLIB_DIR.key(),
+                ConfigConstants.DEFAULT_FLINK_USR_LIB_DIR);
+        return providedUsrLibPath;
+    }
 
-        return getRemoteSharedPaths(
-                configuration,
-                pathStr -> {
-                    final Path path = new Path(pathStr);
-                    return path.getFileSystem(yarnConfiguration).makeQualified(path);
-                });
+    @VisibleForTesting
+    static List<Path> getQualifiedRemoteSharedPaths(
+            List<String> rawPaths, String configOptionKey, YarnConfiguration yarnConfiguration)
+            throws IOException {
+        final List<Path> providedDirs = new ArrayList<>();
+        for (String rawPath : rawPaths) {
+            providedDirs.add(strToPathMapper(rawPath, yarnConfiguration));
+        }
+        return validateRemoteSharedPaths(providedDirs, configOptionKey);
+    }
+
+    private static List<Path> validateRemoteSharedPaths(
```

Comment:
Why do we need a separate method for validation? This just make us run the for loop twice.